### PR TITLE
decode env key and enable loglevels of dataservices

### PR DESCRIPTION
**Pull request by aditya-ranshinge**
## Proposed changes

Added method to decode env key 
Added method to read loglevels of dataservices from environment variables to set loglevels of respective logcomponents created for dataservices.

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## PR Checklist
Please delete options that are not relevant.
- [x] If the changes in this PR are meant for the next release / mainline, this PR targets the "staging" branch.
- [x] My code follows the style guidelines of this project (see: [Contributing guideline](https://github.com/zowe/zlux/blob/master/CONTRIBUTING.md))
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] New and existing unit tests pass locally with my changes
- [ ] video or image is included if visual changes are made
- [ ] Relevant update to CHANGELOG.md
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works, or describe a test method below

### CODE REVIEW

File: c/envService.c [hunk 1]
```diff
@@ -73,6 +73,88 @@ static JsonObject *returnJsonObj(ShortLivedHeap *slh, char buf[]) {
   return envSettingsJsonObject;
 }
 
+static char checkValidHex(const char h)
```
[the remark this answers is not in the document]
Comment (adityadr-joy) [reply c/envService.c]: Removed const qualifier

File: c/envService.c [hunk 2]
```diff
@@ -73,6 +73,88 @@ static JsonObject *returnJsonObj(ShortLivedHeap *slh, char buf[]) {
...
+  } else if (h >= 'a' && h <= 'f') {
+    return h - 'a' + 10;
+  } else if (h >= 'A' && h <= 'F') {
+    return h - 'A' +10;
+  } else {
+    return 0;
+  }
+}
+
+static char convertHexToChar(const char h1, const char h2)
```
[the remark this answers is not in the document]
Comment (adityadr-joy) [reply c/envService.c]: Removed const qualifier

File: c/envService.c [hunk 3]
```diff
@@ -73,6 +73,88 @@ static JsonObject *returnJsonObj(ShortLivedHeap *slh, char buf[]) {
...
+    decodedChar = (hexCheck1 << 4) | hexCheck2;
+    a2e(&decodedChar,1);
+  }
+  return decodedChar;
+}
+
+#define MAX_ENCODED_UNDER_SCORE_COUNT 4
+static void decodeEnvKey(const char *envKey, const size_t keyLen, char *decodedKey)
+{
+  int _count = 0, j = 0, skipPrint = 0;
```
[the remark this answers is not in the document]
Comment (adityadr-joy) [reply c/envService.c]: removed _ from variable name

File: c/envService.c [hunk 4]
```diff
@@ -16,6 +16,8 @@
 #include "alloc.h"
 #include "json.h"
 
+#define RC_CHECK_VALID_HEX_FAILED 100
```
Comment (lchudinov) [inline c/envService.c:19]: `-1` would be better here.

File: c/zss.c [hunk 5]
```diff
@@ -698,37 +698,81 @@ static void installWebPluginDefintionsService(WebPluginListElt *webPlugins, Http
...
-          }
-          property = jsonObjectGetNextProperty(property);
+#define ENV_LOGLEVELS_PREFIX "ZWED_logLevels_"
+static int getLogLevelsFromEnv(JsonObject *envConfig, const char *identifier) {
+  int len = strlen(ENV_LOGLEVELS_PREFIX) + strlen(identifier) + 1;
+  int logLevel = ZOWE_LOG_SEVERE;
+  char envKey[len];
+  memset(envKey,'\0',len);
+  snprintf(envKey, len, "%s%s", ENV_LOGLEVELS_PREFIX, identifier);
+  logLevel = jsonObjectGetNumber(envConfig, envKey);
```
Comment (lchudinov) [inline c/zss.c:708]: `jsonObjectGetNumber` will return `0` when 
* the key not found
* the value is `0`

You should distinguish the cases.

File: c/zss.c [hunk 6]
```diff
@@ -698,37 +698,81 @@ static void installWebPluginDefintionsService(WebPluginListElt *webPlugins, Http
...
+  return logLevel;
```
Comment (lchudinov) [inline c/zss.c:709]: Again, the function returns `0` if the environment variable for the `logLevel`
* is not set
* is set to `0`

We should distinguish the cases.

File: c/zss.c [hunk 7]
```diff
@@ -698,37 +698,81 @@ static void installWebPluginDefintionsService(WebPluginListElt *webPlugins, Http
...
+}
+
+static int checkLoggingVerbosity(JsonObject *serverConfig, JsonObject *envConfig, char *pluginIdentifier) {
+  int logLevel = getLogLevelsFromEnv(envConfig, pluginIdentifier);
```
Comment (lchudinov) [inline c/zss.c:713]: When should distinguish the case when the env is not set versus the case when the env var is set to `0`.

File: c/zss.c [hunk 8]
```diff
@@ -698,37 +698,81 @@ static void installWebPluginDefintionsService(WebPluginListElt *webPlugins, Http
...
+    }
+  }
+  return dataServiceIdentifier;
+}
+
+static void checkAndSetDataServiceLoglevel(JsonObject * serverConfig,
+                                           JsonObject * envConfig,
+                                           char *dataServiceIdentifier, 
+                                           uint64 loggingIdentifier) {
+  int logLevel = getLogLevelsFromEnv(envConfig, dataServiceIdentifier);
```
Comment (lchudinov) [inline c/zss.c:753]: Same as above.

File: c/zss.c (hunk 5 above)
Comment (adityadr-joy) [reply c/zss.c]: Added jsonObjectHasKey() check before jsonObjectGetNumber() call.

File: c/zss.c (hunk 7 above)
Comment (adityadr-joy) [reply c/zss.c:718]: Added jsonObjectHasKey() check in getLogLevelsFromEnv(). So if env is not set then getLogLevelsFromEnv() returns ZOWE_LOG_NA.

File: c/zss.c (hunk 8 above)
Comment (adityadr-joy) [reply c/zss.c:758]: Added jsonObjectHasKey() check in getLogLevelsFromEnv(). So if env is not set then getLogLevelsFromEnv() returns ZOWE_LOG_NA.

File: c/zss.c (hunk 6 above)
Comment (adityadr-joy) [reply c/zss.c:714]: Added jsonObjectHasKey() check before jsonObjectGetNumber() call.

**Review by lchudinov — APPROVED** [review]
Looks good to me now. Thanks!

**Pull request merged**